xmtp_mls/src/groups/mls_sync.rs

```diff
@@ -2964,19 +2937,69 @@ fn get_removed_leaf_nodes(
         .collect()
 }
 
-fn get_and_clear_pending_commit(
+/// Execute a commit-creating operation using a savepoint pattern.
+/// 
+/// This function:
+/// 1. Runs the operation in a transaction savepoint
+/// 2. Extracts the pending commit data
+/// 3. Rolls back the transaction (avoiding the need for clear_pending_commit)
+/// 4. Reloads the group to clear its internal cache
+/// 5. Returns the operation result and the staged commit
+///
+/// This is more reliable than using `clear_pending_commit` because it uses
+/// SQLite's built-in savepoint rollback mechanism.
+pub(super) fn generate_commit_with_rollback<S, R, E, F>(
+    storage: &S,
     openmls_group: &mut OpenMlsGroup,
-    s: &impl XmtpMlsStorageProvider,
-) -> Result<Option<Vec<u8>>, GroupError> {
-    let commit = openmls_group
-        .pending_commit()
-        .as_ref()
-        .map(xmtp_db::db_serialize)
-        .transpose()?;
-    openmls_group.clear_pending_commit(s)?;
-    Ok(commit)
+    operation: F,
+) -> Result<(R, Option<Vec<u8>>), GroupError>
+where
+    S: XmtpMlsStorageProvider,
+    E: Into<GroupError>,
+    F: for<'a> FnOnce(&mut OpenMlsGroup, &XmtpOpenMlsProviderRef<<S::TxQuery as TransactionalKeyStore>::Store<'a>>) -> Result<R, E>,
+{
+    let mut result = None;
+    let mut staged_commit = None;
+    
+    let transaction_result = storage.transaction(|conn| {
+        let key_store = conn.key_store();
+        let provider = XmtpOpenMlsProviderRef::new(&key_store);
+        
+        // Execute the operation (e.g., self_update, update_group_context_extensions, etc.)
+        result = Some(operation(openmls_group, &provider));
+        
+        // Extract the staged commit data before rollback
+        staged_commit = openmls_group
+            .pending_commit()
+            .as_ref()
+            .map(xmtp_db::db_serialize)
+            .transpose()
+            .ok();
+        
+        // Rollback the transaction to avoid persisting the commit
+        Err::<(), StorageError>(StorageError::IntentionalRollback)
+    });
+    
+    // Reload the group to clear its internal cache after rollback
+    openmls_group.reload(storage)?;
+    
+    // Check if the transaction was intentionally rolled back (expected)
+    // or if there was a real error
+    if let Err(e) = transaction_result {
+        if !matches!(e, StorageError::IntentionalRollback) {
+            return Err(e.into());
+        }
+    }
```

Comment:
This should probably be `let Err(e) = transaction_result else {<panic or return error>}`​ and then the check for an error other than rollback can be checked.